test/dynamo/test_autograd_function.py

```diff
@@ -0,0 +1,255 @@
+            grad_bias = grad_output.sum(0)
+
+        return grad_input, grad_weight, grad_bias
+
+
+class ModuleLinear(torch.nn.Module):
+    def forward(self, input, weight, bias=None):
+        return LinearFunction.apply(input, weight, bias)
+
+
+class MaterializingGradFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        ctx.set_materialize_grads(False)
+        return x.clone(), x.clone()
+
+    @staticmethod
+    def backward(ctx, grad_out1, grad_out2):
+        return grad_out1, grad_out2
+
+
+class MaterializingGradModule(torch.nn.Module):
+    def forward(self, x):
+        return MaterializingGradFunction.apply(x)
+
+
+class CustomFuncBwdPrintGraphBreak(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        return torch.add(foo, foo)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        print("graph break!")
+        return grad_output
+
+
+class CustomFuncBwdPrintModule(torch.nn.Module):
+    def forward(self, x):
+        return CustomFuncBwdPrintGraphBreak.apply(x)
+
+
+class CustomFuncStrideBwd(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        return torch.add(foo, foo)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.stride()
+
+
+class CustomFuncStrideModule(torch.nn.Module):
+    def forward(self, x):
+        return CustomFuncStrideBwd.apply(x)
+
+
+class CustomFuncSaveForBwd(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, foo):
+        result = foo + foo
+        result = result + foo
+        ctx.save_for_backward(result)
+        return result
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        (result,) = ctx.saved_tensors
+        return grad_output * math.sqrt(result.numel())
+
+
+class SaveForBwdModule(torch.nn.Module):
+    def forward(self, foo):
+        return CustomFuncSaveForBwd().apply(foo)
+
+
+class AutogradFunctionTests(torch._dynamo.test_case.TestCase):
+    # Sound behaviors, tested for working capture
+    def test_autograd_function_equivalence(self):
+        for grad in [True, False]:
+            for i in range(1, 5):
+                torch._dynamo.reset()
+                model = globals()[f"Module{i}"]()
+                opt_model = torch._dynamo.optimize("eager")(model)
+                self.assertTrue(
+                    torch.allclose(
+                        opt_model(torch.ones(2, 3, requires_grad=grad)),
+                        torch.tensor([2.0], requires_grad=grad),
+                    )
+                )
+
+    def test_autograd_function_has_graph_break(self):
+        for grad in [True, False]:
+            x = torch.randn(10, requires_grad=grad)
+            for model in [Module5(), Module6()]:
+                torch._dynamo.reset()
+                cnts = torch._dynamo.testing.CompileCounter()
+                opt_model = torch._dynamo.optimize(cnts)(model)
```

Comment:
nit: are all of the torch._dynamo.optimize calls just `torch.compile`?

Reply:
Yes but I prefer this.